### ISSUE TITLE
fn start fails for non-standard OSX docker installations

### ISSUE THREAD

**Issue by prologic** [fnproject/fn#779]
```
$ fn start
time="2018-02-18T00:23:22Z" level=info msg="Setting log level to" level=info
time="2018-02-18T00:23:22Z" level=info msg="started tracer" url=
time="2018-02-18T00:23:22Z" level=info msg="datastore dialed" datastore=sqlite3 max_idle_connections=256
time="2018-02-18T00:23:22Z" level=error msg="Could not open BoltDB file for MQ" dir=/app/data error="invalid argument" mq=bolt
time="2018-02-18T00:23:22Z" level=fatal msg="Error during server opt initialization." error="invalid argument"
2018/02/18 00:23:22 error: processed finished with error exit status 1
```

**Comment by skinowski** [comment]
@prologic BoltDB tries to open ./data/fn.mq file under your current directory. (eg. /home/prologic) So I suspect it's unable to read this file, could be a file/dir permission issue. Also what is your os/platform?

**Comment by prologic** [comment]
OMG. This is the same/similar issue to iron-io/functions#631 -- Are we just copying off each other and making the same assumptions?

Can we *not* assume a local Docker engine running on localhost? Please :)

I know what the problem is now; I'm running `fn start` with a remote Docker engine; so `/home/blah` will not exist :/

**Comment by prologic** [comment]
Also bind mounting volumes like this is really silly. In production you would actually use a named volume using the docker volumes sub-system and an appropritate volume driver. Very rarely (*never?*) have I bind mounted volumes for "stateful" services.

**Comment by carimura** [comment]
@prologic fyi Fn forked IronFunctions almost a year ago. It's *much* further along, nearing GA in fact, but will have similar issues as a result of same ancestry.

**Comment by prologic** [comment]
I guess it's not a "blocker" per se; but if you want folks to be able to try fn out quickly; fixing this is a good thing :)

**Comment by prologic** [comment]
Also thanks for the history lesson; good to see code re-use (*even if a fork*).

**Comment by treeder** [comment]
`fn start` is mostly just for quickstart/local development. Anything else and you should use `docker run ...` with whatever flags you want.

**Comment by prologic** [comment]
So what about quickstart on OS X?
Is this well documented/tested?

`fb start` doesn't work for me with
OS X + docker-machine managed engine (virtualbox)

**Comment by denismakogon** [comment]
@prologic can you be more explicit when saying that `fn start` doesn't work for you?

**Comment by prologic** [comment]
Was my paste not clear?


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by denismakogon** [comment]
Okay. I can confirm that `fn start` works perfectly with Docker for Mac Version 17.12.0-ce-mac49 (21995). 

If it seems that `fn start` is faulty for you. You can get customised Fn server by creating an env file (identical for Docker env file) and pass that to the `fn start`, check `fn start --help` for more details.
You can also try the following compose file https://gist.github.com/denismakogon/1188d8b2d44f51167ad5a73f98a6def7 (modified version of the one we have in-tree)

**Comment by prologic** [comment]
> Okay. I can confirm that fn start works perfectly with Docker for Mac
Version 17.12.0-ce-mac49 (21995).

So I'm not using Docker for MAc; but docker-machine managed. Not sure if
there's any significant differences; but probably is.
I think Docker for Mac does some host<->engine filesystem sharing magic.

**Comment by willjstevens** [comment]
**Permissions issue.**

When you run with `sudo` - it works in any directory. Without sudo, it fails.

Try:

`$ sudo fn start`

I looked at the source code printing this error; I think the line above it has too restrictive or demanding permissions on the file. This should be loosened somehow.

db, err := bolt.Open(url.Path, **0655**, &bolt.Options{Timeout: 1 * time.Second})
if err != nil {
log.WithError(err).Errorln("Could not open BoltDB file for MQ")
return nil, err
}

So sudo raises the privileges level. Too bad this bolt.go file is in a different "iron-io" project.

**Comment by xolve** [comment]
Has this been released? I am on Docker version `18.03.0-ce, build 0520e24` and I am using `docker-machine` command to manage docker daemon instead of Docker.app. I start docker as `docker-machine start`. It asks me to set environment variables via output of `docker-machine env` in current shell. If I don't  fn wouldn't find docker daemon and give this error:

```docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
See 'docker run --help'.
2018/04/11 00:12:28 error: processed finished with error exit status 125```

I set the following environment variables:
```export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://192.168.99.100:2376"
export DOCKER_CERT_PATH="/Users/user/.docker/machine/machines/default"
export DOCKER_MACHINE_NAME="default"```

fn start gives me error:

```time="2018-04-11T01:57:03Z" level=info msg="Setting log level to" level=info
time="2018-04-11T01:57:03Z" level=info msg="datastore dialed" datastore=sqlite3 max_idle_connections=256
time="2018-04-11T01:57:03Z" level=error msg="Could not open BoltDB file for MQ" dir=/app/data error="invalid argument" mq=bolt
time="2018-04-11T01:57:03Z" level=fatal msg="Error during server opt initialization." error="invalid argument"
2018/04/11 12:21:03 error: processed finished with error exit status 1```

**Comment by rdallman** [comment]
@xolve yes this has been released, hopefully it fixes this. thanks for the detailed bug report, we will try to test with docker machine, the boltdb error there isn't very helpful and we should try to nail that one down.
